python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1026,10 +1038,25 @@ predicate kwOverflowStoreStep(CfgNode nodeFrom, DictionaryElementContent c, Node
   )
 }
 
+/**
+ * Simulation of flow summary for `json.loads`.
+ */
+predicate jsonLoadsStoreStep(CfgNode nodeFrom, Content c, Node nodeTo) {
+  exists(CallNode call, ControlFlowNode json |
+    json.isAttribute() and
+    json.getNode().(Attribute).getName() = "loads" and
+    json.getNode().(Attribute).getObject().(Name).getId() = "json" and
+    call.getFunction() = json and
+    nodeFrom.asCfgNode() = call.getArg(0) and
+    nodeTo.asCfgNode() = call and
+    c instanceof RecursiveElementContent
+  )
+}
```

Comment:
Recently we discussed whether type-tracking and data-flow would end up recursively depending on each other, or if we could have a strictly layered approach (type-tracking depends on data-flow, but data-flow can't depend on type-tracking)...

My assumption (without thinking about performance implications) has always been that we need type-tracking to be able to define some data-flow things. However, the general assumption seemed to be that we could avoid this. 

I think this is a prime example of needing type-tracking/API graphs to track `json.loads` (like we [currently do](https://github.com/github/codeql/blob/71a6ef5b00c21ab4bdad6380a7bfc65ccbf10e0d/python/ql/src/semmle/python/frameworks/Stdlib.qll#L1058)), so we can define this store step properly.

I guess a discussion in person is better suited to figure out what to do about this

Reply:
Indeed, as I understand it, flow summaries gives you a hook to define load and store steps. As I see it, we would use type trackers to select the function, so it all becomes recursive in the end.

Reply:
So, here's my view on this. Type trackers need two things in order to function: local data flow and call graphs. The call graph is implemented using type tracking, so there's recursion here.

Type trackers _cannot_ depend on global data flow. That would be catastrophic, as any change in flow (e.g. addition of flow steps) would then force a recomputation of everything that depends on type tracking, including the call graph.

So, if we decide to use API graphs to specify things like `json.loads`, then it must be the case that this _does not_ end up affecting type tracking itself (and hence this cannot affect local flow either).

This suggests that at the very least, content flow steps of various kinds must be kept out of the "simple local flow step" core that type tracking depends on. We may need to split up our definition of local flow for this, having one for "local flow for type trackers" and one (most likely a superset) for "local flow for global flow".

Reply:
Totally agree that we cannot let type-trackers depend on global data-flow. However, I don't see the connection between that and the conclusion _then it must be the case that this does not end up affecting type tracking itself (and hence this cannot affect local flow either)_.

(I don't actually know if modeling with flow-summaries ends up affecting local flow or not, to be totally honest)

Reply:
> Totally agree that we cannot let type-trackers depend on global data-flow. However, I don't see the connection between that and the conclusion _then it must be the case that this does not end up affecting type tracking itself (and hence this cannot affect local flow either)_.

I think you may have misread what I wrote. I'm not concluding anything -- I'm stating a prerequisite that must be satisfied if we want to avoid a huge performance impact.

Perhaps this is best explained by considering a few cases. My observation basically boils down to the following question: "What is affected by adding a flow summary?"

#### Case 1: Only global data flow is affected. The set of local flow steps remains unchanged.

In this case, there's no issue, and we can freely use API graphs to specify flow summaries.

#### Case 2: Adding a flow summary affects local (and by extension global) data flow, but can be done "universally" across all queries.

In this case using API graphs to specify flow summaries will introduce recursion between type tracking and the definition of the flow summaries (since the flow summaries use API graphs in their specification, but in turn affect the local flow basis of type tracking).

This could still be okay, since we're only paying the penalty once, but I don't know how realistic the "universality" is.

#### Case 3: Adding a flow summary affects local (and by extension global) data flow, and different queries require different flow summary extensions.

This one is especially bad, since we'll now end up recalculating type trackers and call graphs for each query that extends the meaning of local flow.

To be honest, I don't like case 2 much either. I think we'll have a much more well-behaved system if we can stick to case 1.